src/vs/workbench/contrib/codeActions/browser/codeActionsContribution.ts

```diff
@@ -78,7 +86,15 @@ export class CodeActionsContribution extends Disposable implements IWorkbenchCon
 		const newProperties: IJSONSchemaMap = { ...codeActionsOnSaveDefaultProperties };
 		for (const [sourceAction, props] of this.getSourceActions(codeActionContributions)) {
 			newProperties[sourceAction] = {
-				type: 'boolean',
+				type: 'object',
```

Comment:
Actually maybe to make this easier we can deprecate the existing boolean option in favor of an enum:

```
"never" | "always" | "explicit"
```

This should work better in the settings editor than something like `boolean | string` or using the object like I originally proposed

With this, the existing settings would map:

```
false -> "never"
true -> "explicit"
```